TrafficCapture/trafficCaptureProxyServer/src/main/java/org/opensearch/migrations/trafficcapture/proxyserver/netty/ExpiringSubstitutableItemPool.java

```diff
@@ -0,0 +1,304 @@
+package org.opensearch.migrations.trafficcapture.proxyserver.netty;
+
+import io.netty.channel.EventLoop;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayDeque;
+import java.util.LinkedHashSet;
+import java.util.Queue;
+import java.util.StringJoiner;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+
+/**
+ * This class maintains N items for S seconds.  After S seconds, items are expired as per the
+ * specified expiration callback.  Callers can retrieve items from the cache or built on-demand
+ * if no items are available within the cache that are ready to go.
+ *
+ * This class does not use locking.  Instead, it is assumed that one of these will be created for
+ * each netty event loop.
+ */
+@Slf4j
+public class ExpiringSubstitutableItemPool<F extends Future<U>, U> {
+
+    private static class Entry<F> {
+        Instant timestamp;
+        F future;
+        public Entry(F future) {
+            timestamp = Instant.now();
+            this.future = future;
+        }
+
+        @Override
+        public String toString() {
+            final StringBuilder sb = new StringBuilder("Entry{");
+            sb.append("timestamp=").append(timestamp);
+            sb.append(", value=").append(future);
+            sb.append('}');
+            return sb.toString();
+        }
+    }
+
+    public static class PoolClosedException extends RuntimeException { }
+
+    public static class Stats {
+        @Getter
+        private long nItemsCreated;
+        @Getter
+        private long nExpiredItems;
+        @Getter
+        private long nHotGets; // cache hits
+        @Getter
+        private long nColdGets; // cache misses
+        @Getter
+        Duration totalDurationBuildingItems = Duration.ZERO;
+        @Getter
+        Duration totalWaitTimeForCallers = Duration.ZERO;
+
+        @Override
+        public String toString() {
+            return new StringJoiner(", ", Stats.class.getSimpleName() + "[", "]")
+                    .add("nItemsCreated=" + nItemsCreated)
+                    .add("nHotGets=" + nHotGets)
+                    .add("nColdGets=" + nColdGets)
+                    .add("nExpiredItems=" + nExpiredItems)
+                    .add("avgDurationBuildingItems=" + averageBuildTime())
+                    .add("avgWaitTimeForCallers=" + averageWaitTime())
+                    .toString();
+        }
+
+        public long getTotalGets() {
+            return nHotGets+nColdGets;
+        }
+
+        public Duration averageWaitTime() {
+            if (getTotalGets() == 0) {
+                return Duration.ZERO;
+            }
+            return totalWaitTimeForCallers.dividedBy(getTotalGets());
```

Comment:
We should be cautious with this. My understanding is that if we have a situation where one item was called 1 second after being created, and 1000 other items have been created and never called, we'd report a wait time of 1 second, because wait time is only updated when an item is given to a caller.

I don't think it's a bad metric to have, especially if you contexturalize it with `totalItemsCreated` and `totalExpiredItems`, but we shouldn't rely on it being an accurate gauge on its own.

Reply:
(The same thing can happen with build time if a bunch of futures are just building indefinitely)

Reply:
This is only relevant for the items that were provided to the user.  Hot and cold gets are collectively increased by 1 when getAvailableOrNewItem() is called.  totalItemsCreated() will be incremented each time that we supply a new item, including items that may be expired or those that are still pending.

Please let me know if I've got some other issue in these calculations.